packages/radix-vue/src/shared/useBodyScrollLock.ts

```diff
@@ -32,68 +28,83 @@ export function useBodyScrollLock(initialState?: boolean | undefined) {
-        document.body.style.setProperty('--scrollbar-width', `${verticalScrollbarWidth}px`)
-        document.body.style.overflow = 'hidden'
-      }
-
-      if (isIOS) {
-        stopTouchMoveListener = useEventListener(
-          document,
-          'touchmove',
-          (e: TouchEvent) => {
-            if (e.target !== document.documentElement)
-              return
-
-            if (e.touches.length > 1)
-              return
-            e.preventDefault?.()
-          },
-          { passive: false },
-        )
-      }
-
-      // let dismissibleLayer set previous pointerEvent first
-      nextTick(() => {
-        document.body.style.pointerEvents = 'none'
-        document.body.style.overflow = 'hidden'
-      })
+    if (val === 0) {
+      resetBodyStyle()
+      return
+    }
+
+    if (initialOverflow.value === undefined)
+      initialOverflow.value = document.body.style.overflow
+
+    const verticalScrollbarWidth = window.innerWidth - document.documentElement.clientWidth
+    const defaultConfig = { padding: verticalScrollbarWidth, margin: 0 }
+
+    const config = context.scrollBody?.value
+      ? typeof context.scrollBody.value === 'object'
+        ? defu({
+          padding: context.scrollBody.value.padding === true ? verticalScrollbarWidth : context.scrollBody.value.padding,
+          margin: context.scrollBody.value.margin === true ? verticalScrollbarWidth : context.scrollBody.value.margin,
+        }, defaultConfig)
+        : defaultConfig
+      : ({ padding: 0, margin: 0 })
+
+    if (verticalScrollbarWidth > 0) {
+      document.body.style.paddingRight = `${config.padding}px`
+      document.body.style.marginRight = `${config.margin}px`
+      document.body.style.setProperty('--scrollbar-width', `${verticalScrollbarWidth}px`)
+      document.body.style.overflow = 'hidden'
     }
+
+    if (isIOS) {
+      stopTouchMoveListener = useEventListener(
+        document,
+        'touchmove',
+        (e: TouchEvent) => {
+          if (e.target !== document.documentElement)
+            return
+
+          if (e.touches.length > 1)
+            return
+          e.preventDefault?.()
+        },
+        { passive: false },
+      )
+    }
+
+    // let dismissibleLayer set previous pointerEvent first
+    nextTick(() => {
+      document.body.style.pointerEvents = 'none'
+      document.body.style.overflow = 'hidden'
+    })
   }, { immediate: true })
 
-  onBeforeUnmount(() => {
-    if (initialState)
+  return count
+})
+
+export function useBodyScrollLock(initialState?: boolean | undefined) {
+  const locked = ref(initialState)
+  const stack = useBodyLockStackCount()
+
+  if (initialState)
+    stack.value++
+
+  watch(locked, (value, oldValue) => {
+    if (value === oldValue)
+      return
+
+    if (value === true)
+      stack.value++
+    else
       stack.value--
+  })
 
-    if (stack.value === 0)
-      resetBodyStyle()
+  onBeforeUnmount(() => {
+    if (locked.value)
+      stack.value--
```

Comment:
@CapitaineToinon  seems like reducing the stack size doesn't trigger the watcher, thus not reseting the body style.
Can you look at it?

Reply:
Okay I think I got it working. I'm trying to write some tests to avoid the back and forth but I'm facing some caching issues.

If I disable the cache for `createSharedComposable` then I can't do tests such as ensure once 2 components are unmounted, the body unlocks since the count won't be shared. However if I don't cache `createSharedComposable` then my first tests will affect the following tests. I can just unmount components to ensure it cleans up but if the test fails for some reason and unmount isn't called then the next test will be wrongly affected and fail.

Is there a way with vitest to create a new mock module or something in a beforeEach in order to have the `createSharedComposable` being only shared between vue instances inside a same test and not all of them?

Reply:
Okay let me know how this one looks. Tests are working and the dropdown and popover components work fine on histoire.

Reply:
LGTM @CapitaineToinon !